Modules/posixmodule.c

```diff
@@ -5088,384 +5088,392 @@ os__path_splitroot_impl(PyObject *module, path_t *path)
-    FILE_STAT_BASIC_INFORMATION statInfo;
+    assert(testedType == PY_IFREG || testedType == PY_IFDIR ||
+           testedType == PY_IFLNK || testedType == PY_IFMNT ||
+           testedType == PY_IFLRP || testedType == PY_IFRRP);
 
-    if (!path_converter(s, &_path)) {
-        path_cleanup(&_path);
-        if (PyErr_ExceptionMatches(PyExc_ValueError)) {
-            PyErr_Clear();
-            Py_RETURN_FALSE;
-        }
-        return NULL;
+    DWORD fileDevType = GetFileType(hfile);
+    if ((fileDevType == FILE_TYPE_UNKNOWN && GetLastError()) ||
+        (fileDevType == FILE_TYPE_CHAR) ||
+        (diskOnly && fileDevType != FILE_TYPE_DISK))
+    {
+        return FALSE;
     }
 
-    Py_BEGIN_ALLOW_THREADS
-    if (_path.wide) {
-        if (_Py_GetFileInformationByName(_path.wide, FileStatBasicByNameInfo,
-                                         &statInfo, sizeof(statInfo))) {
-            if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT)) {
-                slow_path = FALSE;
-                result = statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY;
-            } else if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY)) {
-                slow_path = FALSE;
-                result = 0;
-            }
-        } else if (_Py_GetFileInformationByName_ErrorIsTrustworthy(GetLastError())) {
-                    slow_path = FALSE;
-                    result = 0;
+    DWORD attributes, reparseTag;
+    if ((testedType != PY_IFREG && testedType != PY_IFDIR)) {
+        FILE_ATTRIBUTE_TAG_INFO info;
+        if (!GetFileInformationByHandleEx(hfile, FileAttributeTagInfo, &info,
+                                          sizeof(info)))
+        {
+            return FALSE;
         }
+        attributes = info.FileAttributes;
+        reparseTag = info.ReparseTag;
     }
-    if (slow_path) {
-        if (_path.fd != -1) {
-            hfile = _Py_get_osfhandle_noraise(_path.fd);
-            close_file = FALSE;
-        }
-        else {
-            hfile = CreateFileW(_path.wide, FILE_READ_ATTRIBUTES, 0, NULL,
-                                OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    else {
+        FILE_BASIC_INFO info;
+        if (!GetFileInformationByHandleEx(hfile, FileBasicInfo, &info,
+                                          sizeof(info)))
+        {
+            return FALSE;
         }
-        if (hfile != INVALID_HANDLE_VALUE) {
-            if (GetFileInformationByHandleEx(hfile, FileBasicInfo, &info,
-                                            sizeof(info)))
-            {
-                result = info.FileAttributes & FILE_ATTRIBUTE_DIRECTORY;
+        attributes = info.FileAttributes;
+        reparseTag = 0;
+    }
+
+    if (attributes & FILE_ATTRIBUTE_REPARSE_POINT) {
+        return _testReparseTag(reparseTag, testedType);
+    }
+    else if (testedType == PY_IFREG) {
+        return ((fileDevType == FILE_TYPE_DISK) &&
+                !(attributes & FILE_ATTRIBUTE_DIRECTORY));
```

Comment:
As in my previous comment, one can rely on `attributes` being non-zero if it's a file in a filesystem. If no other attribute is set, at least `FILE_ATTRIBUTE_NORMAL` (0x0080) will be set.

```suggestion
        return attributes && !(attributes & FILE_ATTRIBUTE_DIRECTORY);
```